### ISSUE TITLE
Temporary bedgraph from bam_to_bigwig is unsorted and throws an error.

### ISSUE THREAD

**Issue by j-andrews7** [daler/pybedtools#178]
This is less of a real issue and more a suggestion. When trying to use the bam_to_bigwig function, I ran into an issue where the temporary bedgraph from a sorted bam file would end up unsorted and throw an error:

```
/tmp/pybedtools.vi0i6nl9.tmp is not case-sensitive sorted at line 10581393.  Please use "sort -k1,1 -k2,2n" with LC_COLLATE=C,  or bedSort and try again.
```

I'm not entirely sure why it ends up unsorted, but adding a sort to the function remedies this. Adding a catch for the exception and then doing the sort if necessary might be helpful.

``` Python
def bam_to_bigwig(bam, genome, output, scale=False):
    """
    Given a BAM file `bam` and assembly `genome`, create a bigWig file scaled
    such that the values represent scaled reads -- that is, reads per million
    mapped reads.

    (Disable this scaling step with scale=False; in this case values will
    indicate number of reads)

    Assumes that `bedGraphToBigWig` from UCSC tools is installed; see
    http://genome.ucsc.edu/goldenPath/help/bigWig.html for more details on the
    format.
    """
    genome_file = pybedtools.chromsizes_to_file(pybedtools.chromsizes(genome))
    kwargs = dict(bg=True, split=True, g=genome_file)
    if scale:
        readcount = mapped_read_count(bam)
        _scale = 1 / (readcount / 1e6)
        kwargs['scale'] = _scale
    x = pybedtools.BedTool(bam).genome_coverage(**kwargs)
    y = x.sort()  # The fix.
    cmds = [
        'bedGraphToBigWig',
        y.fn,
        genome_file,
        output]
    os.system(' '.join(cmds))
```

**Comment by daler** [comment]
Thanks. I've been hitting similar issues when using bedGraphToBigWig in other contexts. It seems there was a change in recent UCSC tools versions that expects a different kind of sort order than samtools provides when sorting a BAM.

This is a good fix, but I'd prefer to fix the issue outside of this function since running that sort every time will get time-consuming when not required.

**Comment by j-andrews7** [comment]
Yeah, but you could just try to catch the exception and only sort if it throws so that it's not so time-consuming. I was lazy here because I know it's going to be an issue with all my files.

**Comment by daler** [comment]
Very true. I'll leave this issue open until I fix that.

**Comment by radlinsky** [comment]
Hello, is there an update to this bug? I'm still seeing this error...

I'm going to temporarily try out implement j-andrews fix for now...

**Comment by daler** [comment]
Thanks for the reminder, I just added a fix. This issue can happen because a sorted BAM is always sorted by the order of sequences in the header. But if those sequences are not themselves sorted, the resulting bedGraph will not be sorted. Here's a problematic sorted BAM file:

```
@HD     VN:1.3  SO:coordinate
@SQ     SN:chrX         LN:20000000
@SQ     SN:chr2L        LN:23011544
None    0       chrX    11      255     5M      *       0       0       CGACA   IIIII   NM:i:0  NH:i:1
None    16      chrX    71      255     5M      *       0       0       TTCTC   IIIII   NM:i:0  NH:i:1
None    16      chrX    141     255     5M      *       0       0       CACCA   IIIII   NM:i:0  NH:i:1
None    0       chr2L   71      255     5M      *       0       0       GAGAA   IIIII   NM:i:0  NH:i:1
None    0       chr2L   141     255     5M      *       0       0       TGGTG   IIIII   NM:i:0  NH:i:1
None    16      chr2L   151     255     5M      *       0       0       GTTCA   IIIII   NM:i:0  NH:i:1
None    0       chr2L   161     255     5M      *       0       0       GATAA   IIIII   NM:i:0  NH:i:1
None    0       chr2L   211     255     5M      *       0       0       AAATA   IIIII   NM:i:0  NH:i:1
```

In this fix in fcc03db, I'm checking for the text "bedSort" in stderr from bedGraphToBigWig. If it's there, re-sort the bedGraph and try again. I suspect that in most cases `samtools sort` on a BAM with a fixed header will be faster than `bedtools sort` on a large bedGraph, but this should be fine for one-off cases.
